### PR TITLE
[WIP] GPII-3860: Resurrect SA strategy

### PR DESCRIPTION
**Pull request by mrtyler**
Using the binaryauth API is not allowed with ADCs. Sergey pointed me to his previous attempt at using SAs instead (before we decided to go with ADCs instead): #225. This PR merges that branch on top of current master.

Still testing and working things out but wanted to provide a quick look so reviewers can familiarize themselves.

### CODE REVIEW

**Comment by natarajaya** [comment]
Looks good so far!

File: common/modules/gcp-project/main.tf [hunk 1]
```diff
@@ -129,6 +144,7 @@ data "google_iam_policy" "combined" {
     members = [
       "${local.service_accounts}",
       "serviceAccount:${google_service_account.gke_cluster_pod_cert_manager.email}",
+      "serviceAccount:projectowner@${var.project_id}.iam.gserviceaccount.com",
```
Comment (stepanstipl) [inline common/modules/gcp-project/main.tf]: `projectowner@` is duplicate - once coming from `local.service_accounts` and once explicitly listed above

File: common/modules/gcp-project/main.tf [hunk 2]
```diff
@@ -201,6 +217,7 @@ data "google_iam_policy" "combined" {
 
     members = [
       "${local.service_accounts}",
+      "serviceAccount:projectowner@${var.project_id}.iam.gserviceaccount.com",
```
Comment (stepanstipl) [inline common/modules/gcp-project/main.tf]: same  - `projectowner@` is duplicate - once coming from `local.service_accounts` and once explicitly listed above

**Comment by stepanstipl** [comment]
Playing with binary auth, I was able to create attestor via `gcloud` cli with user credentials:
```
$ gcloud beta container binauthz attestors create ${ATTESTOR} \
--attestation-authority-note=${NOTE_ID} \
--attestation-authority-note-project=${PROJECT_ID}

$ gcloud beta container binauthz attestors list
┌───────────────┬───────────────────────────────────────────────────────┬─────────────────┐
│      NAME     │                          NOTE                         │ NUM_PUBLIC_KEYS │
├───────────────┼───────────────────────────────────────────────────────┼─────────────────┤
│ test-attestor │ projects/gpii-gcp-dev-stepan/notes/test-attestor-note │ 0               │
└───────────────┴───────────────────────────────────────────────────────┴─────────────────┘
```

So it seems that the Terraform message (`* google_binary_authorization_attestor.attestor: Error creating Attestor: googleapi: Error 403: Your application has authenticated using end user credentials from the Google Cloud SDK or Google Cloud Shell which are not supported by the binaryauthorization.googleapis.com. We recommend that most server applications use service accounts instead. For more information about service accounts and how to use them in your application, see https://cloud.google.com/docs/authentication/.`) is not accurate (I assume this is the one @mrtyler was referring to when mentioning that the API does not support app default credentials) and this might actually be an issue with Terraform in the first place?


Another way might be, instead of re-introducing this imho quite complex SA strategy, using dedicated svc account just for the bin auth policy. This can be conveniently accomplished by using the `google_service_account_access_token` Terraform data source.

```hcl
data "google_service_account_access_token" "bin-auth" {
  provider = "google-beta"

  # this would be dedicated account just to manage bin auth
  target_service_account = "test-970@gpii-gcp-dev-stepan.iam.gserviceaccount.com"
  scopes = ["cloud-platform"]
  lifetime = "300s"
}

provider "google-beta" {
  alias       = "svc"
  access_token = "${data.google_service_account_access_token.bin-auth.access_token}"
}
```

Then this provider alias can be used in the reource like this:
```hcl
resource "google_binary_authorization_attestor" "attestor" {
  provider = "google-beta.svc"
...
}
```

I have tested this and seems to work just fine.

File: common/modules/gcp-project/main.tf (hunk 1 above)
Comment (mrtyler) [reply common/modules/gcp-project/main.tf]: Yup, thanks.

File: common/modules/gcp-project/main.tf (hunk 2 above)
Comment (mrtyler) [reply common/modules/gcp-project/main.tf]: Yup, thanks.

**Comment by mrtyler** [comment]
Thanks for your reviews @stepanstipl!

Re: ADC vs SA: I think I've captured the main points in this comment: https://issues.gpii.net/browse/GPII-3860?focusedCommentId=39032&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-39032

I agree that SAs introduce more complexity than ADCs -- mostly because every user needs both a named account (for the web interface, and to bootstrap the auth process) and a service account (for terraform operations). Unfortunately, I don't see a better way to proceed.

> How do we handle accounts for "normal" developers?

The same as for admin users: a SA is created in their dev env, both that SA and their user account get admin/owner permissions.

> How do we handle tracking usage of the service accounts to persons?

In normal operation, the code manages this and uses the appropriate user-specific account.

In an attack scenario, key creation and IAM change events go to the audit log. We have some alerts for these events (perhaps we want more to cover the new workflow).

Some time ago, you pointed out that it is easier for an admin/attacker to impersonate another user via SA than to impersonate another user via user accounts. This is true, but I don't see a better way to proceed.

> How will the privilege escalation work for service accounts?

The same as today. The `:grant_(project|org)_admin` tasks will add additional IAM bindings for the user's named account and SA.

> (looks like that service accounts would be always roles/storage.admin and roles/dns.admin on the common project? Is that intentional?)

This branch doesn't change anything about the permissions themselves, only the accounts to which those permissions are assigned (exception: I added some system SA bindings to keep the binaryauth API enabled and working).

**Comment by mrtyler** [comment]
There is still work to do here (mostly fixing the `:grant_*_admin` tasks, some other cleanups) but with this code I can:
* `rake deploy_module[infra/dev/tyler]` (in the test org, assigning admin permissions to a SA manually)
* `rake deploy` a dev-tyler environment (in the test org) including a binauth-enabled k8s cluster

I changed Sergey's email-to-SA naming algorithm. I liked it, but it was hard to implement in TF code. This was the last straw, while trying to take the first 30 characters of a generated SA name:

```
google_service_account.project[4]: substr: 'offset + length' cannot be larger than the length of the string
```

My uglier, less general solution greatly simplifies the TF implementation (there is no way to do complex string operations like `substr()` and `replace()` on lists).

**Comment by stepanstipl** [comment]
> Unfortunately, I don't see a better way to proceed.

I offered a pure TF solution (https://github.com/gpii-ops/gpii-infra/pull/402#issuecomment-500811779) which costs about 10 lines of TF code (well plus creating the SA & it's permissions), requires no modification of Ruby code or anything else, does not create static credentials and works.

As you pointed out we've discussed the pros & cons of using SAs for user-to-service interactions and decided against for many reasons, some listed above. 

Due to the added complexity (the PR is opened since 20th May and still not finished) and negative security implications (it's not enough to just disable GSuite user anymore to block access, over 50+ new service accounts created just across main RTF org projects...), I still can't agree that this is the right direction to go.

**Comment by stepanstipl** [comment]
I've summed up my main objections against moving to SAs in the ticket - https://issues.gpii.net/browse/GPII-3860

**Comment by mrtyler** [comment]
> I offered a pure TF solution (#402 (comment)) which costs about 10 lines of TF code (well plus creating the SA & it's permissions), requires no modification of Ruby code or anything else, does not create static credentials and works.

I responded to this theme [here](https://issues.gpii.net/browse/GPII-3860?focusedCommentId=39510&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-39510) but quickly for the record: I don't think your characterization of your solution is accurate, especially considering the scope of your solution vs the scope of migrating to SAs.

> Due to the added complexity (the PR is opened since 20th May and still not finished)

I think clock time is a poor metric on which to judge the merits of this PR, particularly since that clock time included the Toronto F2F and more than a week of vacation days.

**Comment by mrtyler** [comment]
Anyway, we decided to abandon this approach. See [comment in the ticket](https://issues.gpii.net/browse/GPII-3860?focusedCommentId=39512&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-39512).

**Pull request closed without merging**